[Step2 - 앱 생명주기와 객체 저장] 미션 제출합니다.

ehgud0670 opened the pull request:
> 궁금한 점 1 

```swift
    func searchStockByKind(handler: ((key: Beverage, value: [Beverage])) -> (Void)) {
        let stockByKind = generateStockByKind()
        stockByKind.forEach {
            handler((key: $0.key,
            value: $0.value))
        }
    }
    
    private func generateStockByKind() -> [Beverage: [Beverage]] {
        var stockByKind = [Beverage: [Beverage]]()
        for beverage in stock {
            guard !stockByKind.keys.contains(beverage)
                else {
                    stockByKind[beverage]?.append(beverage)
                    continue
            }
            stockByKind[beverage] = [beverage]
        }
        return stockByKind
    }
```

=> 위 코드(자판기 구조체의 메소드입니다)에서 generateStockByKind는 음료 제품을 딕셔너리 형태로(제품별로) 리턴하는 함수인데
해당 딕셔너리의 value가 배열인 이유는 해당 제품에 대한 재고를 배열에 담기 위한 것이라 합당한 것 같은데, 
key가 Beverage인 것이 좀 아쉽습니다. 왜냐하면 Beverage 인스턴스가 필요한게 아니라 Beverage의 type만 필요할 것 같기 때문입니다. 그래서 Beverage.Type 같은 것도 썼는데 컴파일오류만 나고 잘 안되더라고요.
 딕셔너리의 key를 무엇으로 할지 더 좋은 방법 있으면 지도 부탁드립니다.  

> 궁금한 점 2 

```swift 
    func subtract(from money: Int) -> Int {
        return money - price
    }
```
=> 위 코드는 음료 클래스의 메소드입니다.  이 메소드를 만든 이유가 자판기 객체의 현재 금액에서 음료를 구입하여 음료의 price를 차감하는 경우를 위해서 이고, **해당 음료의 price 프로퍼티를 private으로 두기 위해서 만들었습니다.** 

**그러나, 현재금액에서 음료 값을 차감해야 되는 것은 자판기 객체의 역할인것 같은데 이렇게 음료객체가 대신 해도 될까요?? 궁금합니다.**

ehgud0670 committed:
Style: Models로 이름 바꿈

ehgud0670 committed:
Refactor: Util의 dateFormatter를 extension을 이용해 Date 객체로 옮김
	: 날짜 형식을 문자열로 출력해주는 역할은 Date 객체이므로
	: extension을 이용해 Date객체가 프로퍼티를 갖도록 함

ehgud0670 committed:
Style: 매개변수명 manufacturingDateInfo 으로 수정함
	이유 : 더 적절한 이름이라서
Style: 최대 길이 99자 넘어서 개행시킴(ViewController의 코드에서)

ehgud0670 committed:
Refactor: 확장한 Date 클래스를 Data.swift 파일에 옮김

	이유: 이후에 우유 유통기한에도 확장한 Date를 쓸거라서
	  잘 보이기 위해 새로 파일을 만들었다.

ehgud0670 committed:
Feat: Milk 객체에 유통기한을 나타내는 속성 추가 및 적용

ehgud0670 committed:
Feat: 해당 날짜가 유통기한을 넘는지 안넘는지 입증하는 메소드 추가
	이유 : 사용할때 매개변수로 들어온 날짜가 유통기한을 넘으면
	       해당 우유 객체를 사용하지 않으려고,
		또 유통기한을 넘지 않으면 사용하려고 만들었다.

ehgud0670 committed:
Deleted: 필요없는 테스트 함수들 삭제

ehgud0670 committed:
Feat: 테스트하는 객체에 대한 setup과 tearDown 구현함
    : 각 메소드의 super 메소드도 순서에 맞게 호출함

ehgud0670 committed:
Style: 이전의 이름이 우유인것을 전달안하는 것 같아 이름 마지막에 Milk라고 붙임

ehgud0670 committed:
Feat: 우유 객체의 유통기한 검사하는 테스트 코드 작성함
	* 성공하는 경우, 실패하는 경우 모두 테스트 코드 작성함

ehgud0670 committed:
Feat: 커피객체에 뜨거운 음료인지 확인하는 속성 추가 및 적용

ehgud0670 committed:
Feat: 상품 정보를 출력할때 우유의 추가된 유통기한 속성도 출력되게끔
      description 연산 프로퍼티를 재정의함

	=> 이를 위해 Beverage 의 description 에 @objc 추가함

ehgud0670 committed:
Feat: 상품 정보를 출력할때 커피의 추가된 뜨거운 음료 여부 속성도 출력되게끔
      description 연산 프로퍼티를 재정의함

ehgud0670 committed:
Refactor: 클래스의 모드 속성을 캡슐화를 위해 private 접근 제한자를 둠
	  * var 로 선언된 것도 let으로 수정함

ehgud0670 committed:
Style: hot이 더 적절한 이름인 것 같아 프로퍼티 명 수정함

ehgud0670 committed:
Feat: 뜨거운지 확인하는 메소드 구현 및 테스트 코드 적용

ehgud0670 committed:
Feat: 소다객체에 칼로리를 나타내는 속성 추가 및 적용

ehgud0670 committed:
Style: 프로퍼티 명 hotState로 수정함
       이유: 프로퍼티 명을 명사로 만들기 위해 수정함

ehgud0670 committed:
Refactor: 해당 커피 정보를 더 구체적으로 문자열로 나타내기 위해 수정함

ehgud0670 committed:
Feat: 저칼로리인지 확인하는 메소드 구현 및 테스트 코드 작성

ehgud0670 committed:
Refactor: 저칼로리 최대치를 나타내는 상수 생성 및 적용

ehgud0670 committed:
Style: 실질적으로 칼로리보다 킬로칼로리가 더 적당한 것 같으므로 이름 수정함

ehgud0670 committed:
Feat: 소다 객체의 칼로리 프로퍼티 정보를 음료 정보에 추가하기 위해
      description을 재정의함

ehgud0670 committed:
Feat: 자판기 구조체에 음료 배열 프로퍼티 추가
	* 구조체 생성자를 쓸것이기 때문에 따로 생성자는 구현하지 않음

Delete: 뷰 컨트롤러에서 자판기 구조체 프로퍼티를 삭제함
	이유 : 생성자에 매개변수도 넣어줘야 하고 아직 필요없다는 판단하에 프로퍼티 지움

ehgud0670 committed:
Style: 재고를 나타내는 stock이라는 이름이 더 어울릴 것 같아 재수정함

ehgud0670 committed:
Feat: stock에서 뜨거운 커피들만 리턴하는 hotCoffees 메소드 구현

     * hotBeverages 가 아닌 hotCoffees라는 메소드를 구현한 이유:
	현재 내 음료 객체들 중에 뜨거운 음료를 나타낼 수 있는 음료는
	오로지 커피뿐이다. 그래서 hotCoffees라고 메소드를 구현했고
	이후의 단계에 커피 말고도 뜨거운 음료 종류가 있다면 리팩토링 하려고 한다.

Feat: init 생성자를 만듬

    이유 : private 한 속성을 초기화하려면 따로 init을 구현해야하기 때문이다.

ehgud0670 committed:
Feat: 비교를 위해 Equatable을 채택하고 구현함

ehgud0670 committed:
Refactor: 현재 뜨거운 음료는 커피 뿐이 없으므로 리턴값을 커피 배열로 대체함

ehgud0670 committed:
Feat: 뜨거운 커피들읋 반환하는 hotCoffees에 대한 테스트 코드 작성함

ehgud0670 committed:
Feat: 들어간 돈을 나타내는 money 프로퍼티 생성
     * 생성시 필수가 아니므로 optional 로 선언함

ehgud0670 committed:
Feat: 돈을 넣는 메소드 추가

Refactor: money 프로퍼티를 non-optional Int로 재선언함
이유 : 생각해보니 optional 로 두지 않고 0이라고 해도 될 것 같아서

ehgud0670 committed:
Refactor: 제로(0)에 대한 상수 생성 및 적용
           Quantity enum을 여러 곳에서 쓰기 때문에 파일을 분리함

ehgud0670 committed:
Feat: 현재 잔액을 나타내는 currentMoney 메소드 기능 구현함

ehgud0670 committed:
Feat: 잔액을 나타내는 메소드와 돈을 넣는 메소드에 대한 테스트 코드 작성함

ehgud0670 committed:
Feat: 특정 음료 인스턴스를 재고에 추가하는 메소드 기능 구현

ehgud0670 committed:
Refactor: == 메소드에서 제조일자는 비교 조건에서 제거함
	* 이유 : 음료가 같기 위해 제조일자는 필요없다고 봄

ehgud0670 committed:
Feat: 이름과 용량을 합친 문자열을 리턴하는 메소드 구현함
     => 이후에 딕셔너리의 키로서 사용하려고 메소드 만듬

ehgud0670 committed:
Style: xcode 자동정렬함

ehgud0670 committed:
Refactor: 음료정보를 문자열로 출력하는 부분에서 이름도 추가함
	이유 : 사실 이름에 빼놓았었음. 실수 했었습니다.

ehgud0670 committed:
Deleted: 이름과 용량을 문자열로 출력하는 메소드 삭제
	생각해보니 이후의 딕셔너리 Beverage의 Hashable로 가능할 것 같다.

ehgud0670 committed:
Feat: 1. 전체 상품 재고를 종류별로 리턴하는 메소드 기능 구현함
      2. Beverage 객체가 Hashable을 채택함

	* 키를 Beverage로 둔 이유
	  : 다른 별도의 타입으로 키를 두지 않고 Beverage 만으로 키를 구성할 수 있다고 생각함
	  이를 위해 Beverage 에 Hashable을 채택함

        * Beverage 의 Hashable의 hash 함수에 brand, volume, price, name의 hashValue를 combine에 매개변수로 넘긴 이유
	  : brand와 price를 넘긴 이유는 한번 더 같은 제품인지를 확인하기 위한 용도이고
	  : volume과 name를 넘긴 이유는 아래 예시처럼 같은 이름이지만 용량에 따라 다른 제품으로 구분되기 때문이다.

	  1. 허쉬초콜릿 드링크 쿠키앤크림 190ml , 230ml
	  2. 펩시 다이어트 콜라 240ml, 250ml, 591ml
	  3. 칸타타 프리미엄 라떼 175ml, 275ml, 390ml, 900ml

ehgud0670 committed:
Refactor: combine에 name과 volume만 넘김
	이유 1: 사실 combine 에 넣어야 할것은 hashValue가 아닌
	       Hashable 을 채택한 객체이다. 잘못알았다.
	이유 2: name과 volume만 combine에 넣어도 음료간에 구분이 되서
	        name + String(volume) 형태의 문자열을 넘겨주었다.

ehgud0670 committed:
Feat: 재고를 종류별로 딕셔너리로 반환하는 메소드를 테스트하는 코드 구현함
      리턴된 딕셔너리를 for문을 이용해
      각 제품의 개수가 몇개인지 비교함으로써 테스트 코드를 구현함

ehgud0670 committed:
Refactor: guard 문 필요없어서 지우고 비교해서 retrun 만 하도록 재수정함

ehgud0670 committed:
Refactor : guard 문 조건이 틀려서 수정함

ehgud0670 committed:
Feat: 현재 금액으로 구매 가능한 음료수 목록을 리턴하는 메소드 및 테스트 코드 구현

	리턴값을 딕셔너리로 한 이유
	: 구매가능한 목록을 현재 재고 중에 구매가능한 객체 모두라고 생각했기 때문에
	  제품별로 배열에 담아 모두 리턴하였다.

ehgud0670 committed:
Feat: 유통기한이 지난 우유들을 반환하는 메소드 기능 구현함
	이유 : 유통기한은 우유만이 가진 속성이기 때문에 특별히 우유 배열를 리턴타입으로 두었다.
	       이 중에 유통기한이 지난 우유들만 리턴하도록 하였다.

Style: 메소드 성격에 따라 메소드 재위치시킴

ehgud0670 committed:
Refactor: guard문의 조건이 틀렸기 때문에 코드 수정함

ehgud0670 committed:
Feat: 유통기한 지난 우유들을 반환하는 메소드에 대한 테스트 코드 구현함

ehgud0670 committed:
Refactor: equatable의 조건이 이름과 용량만 같아도 된다고 판단하여 코드 수정함

ehgud0670 committed:
Feat: 음료수를 파는 메소드 구현 ( 사용자 입장에서 음료수를 구매하는 메소드 구현)

      * 음료 객체의 price를 private에서 internal 제한자로 둔 이유
	: 상위 모듈의 money에서 해당 음료의 price를 빼려면 어쩔수 없이 price 값을 리턴받아야 한다.
	  따라서 internal 제한자로 두어 상위모듈에서 접근할 수 있게끔 하였다.

      * 음료수를 팔때 재고에서 해당 음료수를 제거하고,
	현재 금액에서 해당 음료 가격을 차감하도록 하였다.
	또 리턴값은 해당 인스턴스가 아닌 동등한 음료 객체를 리턴하도록
	== (Equatable) 메소드를 사용하였다.

ehgud0670 committed:
Feat: 음료수를 파는 메소드에 대한 테스트 코드 구현함

ehgud0670 committed:
Style: 메소드들 재위치시킴
	기준: 무언가를 리턴하는 메소드들은 아래에 위치시킴

ehgud0670 committed:
Style: 함수명 Sellable로 바꿈
	이유: 자판기입장에선 팔 수 있는 음료들이라서 함수 이름 수정함
	     (일관성을 지키기 위해)

ehgud0670 committed:
Feat: 시작부터 구매상품 이력을 배열로 리턴하는 메소드, 테스트 코드 구현함

	구매상품 이력을 프로퍼티로 두었다.

ehgud0670 committed:
Style: 더 적절하게 함수명 및 매개변수 명 수정함

ehgud0670 committed:
Refactor: 해당 메소드는 반환값이 배열이라
	  클로저로도 구현가능하므로 리팩토링함

ehgud0670 committed:
Refactor: 해당 함수는 리턴값이 배열이라 클로저를 활용할 수 있으므로 리팩토링함

ehgud0670 committed:
Refactor: 해당함수는 리턴값이 배열이라 클로저를 활용할 수 있으므로 리팩토링함

ehgud0670 committed:
Refactor: 리턴값이 딕셔너리 이므로 클로저를 활용할 수 있으므로 리팩토링함
	  클로저의 매개변수로 key: Beverage, value: [Beverage] 의 형태인 튜플을 둠

ehgud0670 committed:
Refactor: 해당 함수는 리턴값이 딕셔너리이므로 클로저를 활용할 수 있어서 리팩토링함
	 * 클로저의 매개변수는 튜플 타입(key: Beverage, value: [Beverage])이다.

Refactor: sell 메소드 discardableResult로 어노테이션함
	  이유: 리턴된 음료를 사용하지 않는 경우도 있어서 어노테이션하였다.

ehgud0670 committed:
Refactor: price를 private 하게 설정 및 적용

	이유: 생각해보니 이외의 모든 프로퍼티가 private 이므로
	price도 다시 private 하게 둠

ehgud0670 committed:
Feat: 상품을 추가하는 메소드에 대한 테스트 코드 작성함

ehgud0670 committed:
Style: 함수명 더 적절하게 변경

ehgud0670 committed:
Refactor: for문 대신에 forEach문을 사용함

godrm commented:
> > 궁금한 점 1
> 
> key가 Beverage인 것이 좀 아쉽습니다. 왜냐하면 Beverage 인스턴스가 필요한게 아니라 Beverage의 type만 필요할 것 같기 때문입니다. 그래서 Beverage.Type 같은 것도 썼는데 컴파일오류만 나고 잘 안되더라고요.
> 딕셔너리의 key를 무엇으로 할지 더 좋은 방법 있으면 지도 부탁드립니다.

Meta Type 이라는게 있습니다. 
타입 자체를 값처럼 비교할 수 있습니다. 
ObjectIndentifider 와 type(of:)를 찾아보세요
> 
> > 궁금한 점 2
> 
> **그러나, 현재금액에서 음료 값을 차감해야 되는 것은 자판기 객체의 역할인것 같은데 이렇게 음료객체가 대신 해도 될까요?? 궁금합니다.**

이상하네요. 하위 모듈이 상위 모듈 값를 바꾸는 건 이상하죠.
자판기에서 하거나 음료수가 아니라 다른 객체가 할 수는 있겠죠. 
잔고 / 금액만 관리하는 곳이 있다면요

ehgud0670 commented:
넵, 피드백 감사합니다. 👍
혹시 다른 피드백은 없으신가요?
지금 수정 중에 있는데 다른 피드백 있으시면 더 수정해보려고 합니다.

godrm commented:
> 넵, 피드백 감사합니다. 👍
> 혹시 다른 피드백은 없으신가요?
> 지금 수정 중에 있는데 다른 피드백 있으시면 더 수정해보려고 합니다.

ㅎㅎ 질문만 보고 아직 코드를 못 봤어요. 지금 보는 중입니다.

ehgud0670 commented:
> ㅎㅎ 질문만 보고 아직 코드를 못 봤어요. 지금 보는 중입니다.

아 넵 ㅎㅎ 감사합니당

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverage.swift:
```diff
@@ -0,0 +1,62 @@
+    
+    func subtract(from money: Int) -> Int {
+        return money - price
+    }
```
앞에도 언급했지만 이 부분은 어색하네요. 자기 속성을 바꾸는 건 괜찮지만 다른 속성을 바꾸는 건 이상합니다.

ehgud0670 replied:
그러면 음료수 객체의 price 속성을 private(set)으로 두고 자판기 객체의 money에서 price를 빼도록 하는 건 어떻게 생각하시나요?
제 생각에는 price 속성이 상수에 Int타입이니깐 접근 제한자를 private(set)이나 intetnal 로 바꿔도 될꺼 같습니다만... 
JK의 의견이 궁금합니다!

godrm commented on VendingMachineApp/VendingMachineApp/Models/Cantata.swift:
```diff
@@ -0,0 +1,22 @@
+import Foundation
+
+class Cantata: Coffee {
+    
```
하위 클래스를 만드는 다는 것은 부모 클래스에 없는 하위 속성을 갖거나 다른 기능을 갖고 있어야 의미가 있습니다.
그렇지 않으면 상위 클래스로 써도 되는거죠. 자판기 미션에서는 하위 클래스를 만들 때 더 구체적인 의미를 부여하는 연습을 해야 합니다. 개선해보세요.

ehgud0670 replied:
아하 넵 알겠습니다. 피드백 반영하겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Models/Coffee.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+    
+    func isHot() -> Bool {
+        return hotState
```
데이터 모델을 만들 때 이런 데이터 양을 숫자 형태로 담도록 설계하는 게 좋습니다.
Bool 타입은 기준 값에 따라 판단한 결과로 쓰는게 더 좋습니다. 그렇지 않으면 기준에 따라, 혹은 결과에 따라 모델이 확장하기 어려워 유연하지 않은 구조가 됩니다.

ehgud0670 replied:
넵, 좋은 말씀 감사합니다. 확장하기 어려운 Bool 타입 대신 Int타입으로 만들어서 구조가 유연하도록 만들어보겠습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Models/Date.swift:
```diff
@@ -1,20 +1,23 @@
-
+    
+    var description: String {
+        return Date.dateFormatter.string(from: self)
```
description은 일반적인 문자열을 표현하는 형식이니까 본인 의도를 표현하는 메소드를 만드는 게 어떨까요?
모든 description이 이런 규칙으로 동작해야 하는 경우에만 이렇게 표현하는 게 좋겠습니다.

ehgud0670 replied:
아, description이 일반적인 문자열을 표현하는 형식이므로 위와 같이 Date.dateFormatter.string(from: self)를 리턴하는 메소드는 description라는 이름이 적절하지 않은 거군요. 넵, 함수명을 변경해서 제 의도를 표현하도록 하겠습니다. 감사합니다.

ehgud0670 replied:
> 본인 의도를 표현하는 메소드를 만드는 게 어떨까요?

`toStringYearMonthDay`으로 표현하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Models/Soda.swift:
```diff
@@ -0,0 +1,38 @@
+    }
+    
+    func isLowCalorie() -> Bool {
+        guard kiloCalorie <= Quantity.maxLowKiloCalorie
```
이렇게 판단하는 경우는 기준 값을 매개변수로 전달하면 더 유연한 구조가 됩니다. 

guard 문법을 썼는데 if 와 guard를 선택하는 본인의 기준은 무엇인가요?
조건을 비교만 해도 되는거면 `return kiloCalorie <= Quantity.maxLowKiloCalorie` 로 충분할 것 같아서요. 
코드를 작성했으면 그 의도를 명확하게 하는 기준이 있어야 합니다.

ehgud0670 replied:
넵, 맞습니다... 사실 강박적으로 어떤 조건에 대해서 guard문을 찾게 되는 것 같아요. 
사실 이거 말고 다른 것들도 guard문을 썻다가 다시 수정했거든요. 
바로 수정하겠습니다. 코드를 작성하는 명확한 기준을 가져보겠습니다. 감사합니다~!

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -0,0 +1,140 @@
+        money += insertedMoney
+    }
+    
+    func currentMoney() -> Int {
```
가격, 잔액에 대한 Money 객체를 만들어보세요. Int 보다 더 구체적으로 돈과 관련된 기능을 담당해보세요.

ehgud0670 replied:
넵. Money 객체를 만들어서 변화에 대해 유연해질 수 있도록 하고 가격, 잔액에 대한 책임을 가지도록 하겠습니다. 감사합니다.

ehgud0670 replied:
* 네 반영하였습니다. 

> 궁금한 점 1 
 
```swift
class Money {
    
    private var balance = Quantity.zero
    var currentBalance: Int {
        return balance
    }
    
    func subtract(price: Int) {
        balance -= price
    }
    
    func plus(money: Int) {
        balance += money
    }
    
    func isEnoughToBuy(price: Int) -> Bool {
        return balance >= price
    }

}

struct VendingMachine {
    
    private var stock: [Beverage]
    private var soldBeverages = [Beverage]()
    private let money = Money()
    ///이하 생략
```

저는 위의 코드처럼 자판기 객체에서 머니 객체를 상수로 선언하고 싶어서 해당 객체를 구조체가 아닌 클래스로 하였습니다. 하지만 모델 객체는 일단 구조체로 하는 게 낫다고 들었는데 이렇게 상수로 선언하고 싶다는 이유로 클래스로 만들어도 될까요? 궁금합니다. 

> 궁금한 점 2 

머니 객체에서 음료 가격을 차감하기 위해 저는 음료의 price 를 private에서 internal 제한자로 바꾸었습니다. 
이래도 될까요? 일단 저는 price가 Int 타입에 Iet으로 선언되어있어서 internal로 바꾸어도 상관없을 것 같아서 바꾸긴 했습니다. 아직 push 안했는데 push 하면 의견 부탁드려요 ㅎㅎ

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    mutating private func addFromSoldBeverages(to beverage: Beverage) {
+        soldBeverages.append(beverage)
```
함수 이름이 맥락을 잘 모르겠어요. SoldBeverages에서 beverage로 추가하는 건가요? 
구현 내용이랑 반대인 것 같은데요?

ehgud0670 replied:
넵 ㅋㅋ 제가 작문을 잘못했네요.. ㅠ 함수명 수정하겠습니다. 감사합니다.

ehgud0670 replied:
add 함수를 일반화할 수 있게 되었습니다. 👍
 
```swift
    mutating func addToStock(beverage: Beverage) {
        add(beverage: beverage, to: stock)
    }
    
    mutating func addToSoldBeverages(beverage: Beverage) {
        add(beverage: beverage, to: soldBeverages)
    }
    
    mutating private func add(beverage: Beverage, to beverages: [Beverage]) {
        var beverages = beverages
        beverages.append(beverage)
    }
```

ehgud0670 replied:
`addFromSoldBeverages(to beverage: Beverage)` 에서 
`addToSoldBeverages(beverage: Beverage)`로 바꾸었습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -0,0 +1,140 @@
+        return nil
+    }
+    
+    mutating private func subtractFromStock(to wantedBeverage: Beverage) -> Beverage? {
```
이 함수도 읽어보면 어색합니다. Stock에서 원하는 음료수로 빼는 건가요? 
이미 Beverage가 있는데 그걸 return 하는걸까요? 입력과 출력이 같은건가요?

ehgud0670 replied:
네, 좀 이상합니다. 
subtractFromStock 메소드는 재고에서 음료수를 빼주기만 해야 하는데 리턴까지 하니 어색합니다. 

따라서 저는 sell 메소드에서 매개변수로 음료 인스턴스가 들어오니 그것이 있는지 subtractFromStock 메소드로 재고에 있는지 먼저 확인하고(있으면 true를 리턴하고 없으면 false를 반환함) 재고에서 뺐습니다. 

이후 sell메소드에서는 해당 음료 인스턴스를 리턴하도록 하였습니다. 
그 이유는 매개변수로만으로는 해당 음료가 재고에 있는지 알 수 없어서 
재고에 있는지 확인 후 동일한 인스턴스를 리턴하는 것이 합당하다고 생각하기 때문입니다. (있으면 음료 반환, 없으면 nil 반환)

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -0,0 +1,140 @@
+                    continue
+            }
+            handler(milkPassedExpirationDate)
+        }
```
오잉 음료수 종류별로 클로저를 다 만들어주는 건가요? 비교하는 조건도 closure로 받을 수는 없을까요?

ehgud0670 replied:
음료수 종류별로 클로저를 다 만들면 비교하기 어려워서 그러신거죠?

ehgud0670 replied:
리턴값이 배열이 아니고 딕셔너리인 경우 어떻게 클로저로 대체할 수 있나요??
고민해봤는데 잘 모르겠습니다... 알려주세요 ㅜㅜ

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -0,0 +1,140 @@
+    }
+    
+    mutating private func subtractFromMoney(to beverage: Beverage) {
+        guard beverage.isBuyable(money: money)
```
이 함수는 돈에서 음료수를 빼야하는 것처럼 보입니다.

ehgud0670 replied:
넵, 음료의 price를 빼도록 바꾸었습니다. 감사합니다.

godrm reviewed (commented):
클래스 상속 관계를 만들 때 주의해야 할 사항을 남겼습니다.
함수 이름은 너무 구현 내용을 드러내려고 할 필요가 없습니다. 
내부 동작이 어떻게 구현되어 있던지 함수 이름(혹은 시그니처)만 보고, 입력에 따라 원하는 결과를 기대할 수 있으면 됩니다. 
다음 피드백을 확인해보세요.

ehgud0670 committed:
Refactor: guard문 지우고 현재 코드로 수정함
	  이유 : 이 함수에서 guard문을 사용하면 오히려 직관성이 떨어진다.
	         그냥 비교한 결과를 Bool값으로 리턴하는 현재 코드가 훨씬 적절하다.

ehgud0670 committed:
Style: 속성명을 toStringYearMonthDay로 바꿈
       이유 : description은 일반적인 문자열을 나타내는 연산 프로퍼티로
	      Date.dateFormatter.string(from: self)을 반환하는 이 연산 프로퍼티는 적절하지 못하다
	      따라서 더 적절한 이름인 toStringYearMonthDay로 바꾸었다.

ehgud0670 committed:
Style: addFromSoldBeverages(to beverage: Beverage) 에서 addToSoldBeverages(beverage: Beverage)로 바꿈
	=> 이전에 작문을 잘못한 거라서 변경함.

Refactor: 재고에서 해당음료를 제거만하고 음료는 반환하지 않고 실패했는지 성공했는지만을 나타내는 Bool값을 리턴하도록 함
	  이유: 사실 이전의 코드는 입력과 출력이 동등하게(Equaltly) 같은 거라서 크게 의미가 없었다. 따라서 재고에서 음료를 빼주도록만 했다.

	* 하지만 sell메소드의 리턴값은 Beverage?로 유지하였다. 그 이유는 매개변수로 넘겨진 음료가 없을 수도 있는 상황을 대비하기 위해서이다.
	  따라서 해당 음료에 대해 동등하게(Equaltly) 같은 음료가 재고에 있다면 리턴하도록 하였다.

ehgud0670 committed:
Refactor: 동등성(==)이 아닌 동일성(===)를 비교하도록 코드 수정함
	이유: sell의 매개변수는 구체적인 음료 인스턴스이다. sell이 리턴해야 할 것도
	      동일한 매개변수 음료 인스턴스를 그대로 리턴해야한다고 생각한다.
	      따라서 같은 인스턴스가 있는 지 확인하고 재고에서 빼기 위해
		동등성이 아닌 동일성을 비교하였다.

ehgud0670 committed:
Refactor: 함수를 잘 못 짜서 수정함, 접근 제한자를 private으로 둠

ehgud0670 committed:
Refactor: 온도 데이터 형태를 수치화 할 수 있도록 Int 타입으로 변경
Style: 섭씨인 것을 잘 표현하기 위해 속성 명을 celsius로 변경함

	이유 : 데이터 모델을 만들때는 데이터 양을 숫자 형태로 담는 것이 좋다.
	       데이터 모델을 타입 Bool로 하는 것은 기준에 따라 혹은 결과에 따라
		확장하기 어려운 유연하지 않은 구조이기 때문이다.

	       Bool 타입은 기준 값에 따라 판단한 결과로만 쓰는 것이 좋다.

ehgud0670 committed:
Refactor: 매개변수를 음료 객체에서 음료의 price로 변경
	이유 : 이전의 함수는 잔액에서 음료를 빼는 것처럼 보였다. ( 문맥상으로도 어색하다 )
	       잔액에서 돈을 빼는 함수의 경우는 음료를 알 필요가 없고 음료 가격만 알면 된다.
	       따라서 음료의 가격만을 넘겨줘야 하는 코드로 바꾸었다.

ehgud0670 committed:
Style: 클래스 이름으로 Balance가 더 어울려 클래스 명 변경함

ehgud0670 committed:
Feat: 칸타타에 우유 퍼센트, 설탕 퍼센트를 나타내는 속성 및 확인하는
	메소드 추가함

ehgud0670 committed:
Style: 속성명들을 더 적절하게 바꿈

ehgud0670 committed:
Feat: 카카오 함유 비율에 대한 속성 및 메소드 추가

ehgud0670 committed:
Feat: 펩시의 패키지 재질에 대한 속성 추가

ehgud0670 committed:
Feat: 펩시, 허쉬초콜릿드링크의 음료 정보를 문자열로 출력하는 기능 추가

ehgud0670 committed:
Feat: 해당 펩시 음료가 깨지기 쉬운지 아닌지 확인하는 메소드를 추가함
	=> 패키지 재질이 유리이면 true, 아니면 false로 설정함

ehgud0670 committed:
Style: 폴더 Beverages 만들고 음료 객체들 옮김

ehgud0670 committed:
Refactor: Balance 보단 Cashier 가 더 적절해 클래스명 변경함

ehgud0670 committed:
Refactor: 본래 자판기 앱에 있던 메소드들을 Cashier로 옮김
	이유: 이제 Cashier의 책임이기 때문이다.

ehgud0670 committed:
Refactor: guard문을 이용하고 지금 잔액이 음료를 살수 있는지 확인하는 메소드 구현함

ehgud0670 committed:
Refactor: 딕셔너리의 키를 String으로 하고 "\(type(of: Beverage))"를 키에 넣음
1

ehgud0670 committed:
Refactor : 고차함수인 filter를 사용하도록 코드 변경함

ehgud0670 committed:
Refactor: 음료 배열이 아닌 음료를 클로저 매개변수로 수정함

ehgud0670 committed:
Style: 파일들 위치옮김

ehgud0670 commented:
피드백 모두 반영하였습니다.

godrm commented:
클래스 속성 중에 enum 타입으로 선언하는 속성은 잘 변경하지 않는 경우에만 하는 게 좋습니다. 
enum 타입으로 선언한 경우 변경에 유연하지 않고 경직되는 경우가 많습니다. enum에 새로운 case를 추가하면 고쳐야 하는 곳이 여러 곳이라면 나쁜 냄새가 난다고 생각하세요. 참고하세요.
